internal/engine/compiler/impl_amd64.go

```diff
@@ -4487,220 +4443,122 @@ func (c *amd64Compiler) allocateRegister(t registerType) (reg asm.Register, err
-		targetFunctionAddressRegister = tmpRegister
+		c.assembler.CompileRegisterToRegister(amd64.MOVQ, functionAddressRegister, tmpRegister)
+		functionAddressRegister = tmpRegister
 	}
 
-	// Also, we have to put the target function's *wasm.ModuleInstance into amd64CallingConventionModuleInstanceAddressRegister.
-	c.assembler.CompileMemoryToRegister(amd64.MOVQ, targetFunctionAddressRegister, functionModuleInstanceAddressOffset,
-		amd64CallingConventionModuleInstanceAddressRegister)
+	// Also, we have to put the target function's *wasm.ModuleInstance into amd64CallingConventionDestinationFunctionModuleInstanceAddressRegister.
+	c.assembler.CompileMemoryToRegister(amd64.MOVQ, functionAddressRegister, functionModuleInstanceAddressOffset,
+		amd64CallingConventionDestinationFunctionModuleInstanceAddressRegister)
 
 	// And jump into the initial address of the target function.
-	c.assembler.CompileJumpToMemory(amd64.JMP, targetFunctionAddressRegister, functionCodeInitialAddressOffset)
+	c.assembler.CompileJumpToMemory(amd64.JMP, functionAddressRegister, functionCodeInitialAddressOffset)
 
 	// All the registers used are temporary, so we mark them unused.
-	c.locationStack.markRegisterUnused(freeRegs...)
+	c.locationStack.markRegisterUnused(tmpRegister, functionAddressRegister)
 
 	// On the function return, we have to initialize the state.
-	// This could be reached after returnFunction(), so callEngine.valueStackContext.stackBasePointer
-	// and callEngine.moduleContext.moduleInstanceAddress are changed (See comments in returnFunction()).
-	// Therefore we have to initialize the state according to these changes.
 	if err := c.compileModuleContextInitialization(); err != nil {
 		return err
 	}
 
-	// Due to the change to callEngine.valueStackContext.stackBasePointer.
+	// Due to the change to callEngine.stackContext.stackBasePointer.
 	c.compileReservedStackBasePointerInitialization()
 
 	// Due to the change to callEngine.moduleContext.moduleInstanceAddress as that might result in
 	// the memory instance manipulation.
 	c.compileReservedMemoryPointerInitialization()
+
+	// We consumed the function parameters, the call frame stack and reserved slots during the call.
+	c.locationStack.sp = uint64(nextStackBasePointerOffset)
+
+	// Now the function results are pushed by the call.
+	for _, t := range functype.Results {
+		loc := c.locationStack.pushRuntimeValueLocationOnStack()
+		switch t {
+		case wasm.ValueTypeI32:
+			loc.valueType = runtimeValueTypeI32
+		case wasm.ValueTypeI64, wasm.ValueTypeFuncref, wasm.ValueTypeExternref:
+			loc.valueType = runtimeValueTypeI64
+		case wasm.ValueTypeF32:
+			loc.valueType = runtimeValueTypeF32
+		case wasm.ValueTypeF64:
+			loc.valueType = runtimeValueTypeF64
+		case wasm.ValueTypeV128:
+			loc.valueType = runtimeValueTypeV128Lo
+			hi := c.locationStack.pushRuntimeValueLocationOnStack()
+			hi.valueType = runtimeValueTypeV128Hi
+		}
```

Comment:
check here and the rest of the diff switches that don't include ref types. add case: for them with a comment why we drop out, or panic with TODO or "unexpected" if there should never be any